pip/pip-423.md

```diff
@@ -0,0 +1,255 @@
+                cursor.getNumberOfEntriesInBacklog(false));
+    }
+
+    if (Subscription.isCumulativeAckMode(getType())) {
+        return CompletableFuture.failedFuture(new NotAllowedException("Unsupported subscription type."));
+    }
+
+    List<Position> positions = new ArrayList<>();
+    for (Map.Entry<String, String> entry : messageIds.entrySet()) {
+        try {
+            long ledgerId = Long.parseLong(entry.getKey());
+            long entryId = Long.parseLong(entry.getValue());
+            Position position = PositionFactory.create(ledgerId, entryId);
+            positions.add(position);
+        } catch (Exception e) {
+            return CompletableFuture.failedFuture(new NotAllowedException("Invalid message ID."));
+        }
+    }
+
+    Map<String, Long> properties = Collections.emptyMap();
+    acknowledgeMessage(positions, AckType.Individual, properties);
+
+    return CompletableFuture.completedFuture(null);
+}
+```
+
+3.  **Admin API Logic:**
+    The `PersistentTopicsBase` class is updated to handle the overloaded `skipMessages` request. When `numMessages` is 0 and the `messageIds` map is not empty, it routes the request to the new `subscription.skipMessages(Map)` method.
+
+```java
+// in pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java
+protected void internalSkipMessages(AsyncResponse asyncResponse, String subName, int numMessages,
+                                    boolean authoritative, Map<String, String> messageIds) {
+    // ...
+    // In the implementation logic for the topic:
+    // ...
+                 if (!messageIds.isEmpty() && numMessages == 0) {
+                     return sub.skipMessages(messageIds).thenAccept(unused -> {
+                                 log.info("[{}] Skipped messages on {} {}", clientAppId(), topicName, subName);
+                                 asyncResponse.resume(Response.noContent().build());
+                             }
+                     );
+                 }
+                  return sub.skipMessages(numMessages).thenAccept(unused -> {
+                      log.info("[{}] Skipped {} messages on {} {}", clientAppId(), numMessages,
+                              topicName, subName);
+                      // ...
+                  });
+    // ...
+}
+```
+
+## Public-facing Changes
+The existing `skipMessages` API is modified to accept a POST body containing message IDs.
+
+### Public API
+The REST endpoint for skipping messages is updated. To skip by message ID, a client must send a `POST` request with `numMessages` set to `0` in the path and provide a map of message IDs in the JSON body.
+
+*   **Path (v2):** `POST /admin/v2/persistent/{tenant}/{namespace}/{topic}/subscription/{subName}/skip/{numMessages}`
+*   **Path (v1):** `POST /admin/v1/persistent/{property}/{cluster}/{namespace}/{topic}/subscription/{subName}/skip/{numMessages}`
```

Comment:
Actually we can add a new path e.g. skipByMessageIds to avoid the confusing 0 for numMessages is required here

Reply:
@codelipenghui Already modified, are there any other areas that need to be adjusted?